prototypes/coin-selection/CoinSelection.hs

```diff
@@ -0,0 +1,278 @@
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE PartialTypeSignatures #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# OPTIONS_GHC -fdefer-typed-holes #-}
+
+-- |
+-- Copyright: © 2018-2022 IOHK
+-- License: Apache-2.0
+-- Stability: experimental
+--
+-- Mock-up of a multi-asset coin selection interface for UTxO blockchains.
+--
+-- The goal is to come up with the simplest possible API which works for a
+-- Cardano wallet, without committing to concrete types and implementations.
+
+module CoinSelection
+    ( -- * Requirements
+      --
+      -- $req
+
+      -- * Coin selection
+      --
+      -- $naming
+      xyzzy
+    , Selector (..)
+    , Change (..)
+    , SelectionError (..)
+    -- * The transaction type
+    , TxBody (..)
+    , Value
+    , TxOut
+    , UTxO
+    -- ** Information about transactions
+    , Params (..)
+    , AddressOwner (..)
+    ) where
+
+import Prelude
+
+import GHC.Generics (Generic)
+import Data.List.NonEmpty (NonEmpty(..))
+import Data.Map (Map)
+
+-- $req
+-- This is what we need the coin selection to be capable of.
+--
+-- __Use cases__
+--
+--  * Payment
+--
+--  * Join stake pool
+--
+--  * Spend 'Value' from reward account
+--
+--  * Configurable change generation
+--
+--  * Wallet v'Sweep'
+--
+--  * Active UTxO management
+--
+--        * Coalesce UTxO
+--        * Create /n/ new UTxO
+--
+--  * Emergent UTxO management => Add a change output to input tx
+--
+-- __Constraints__
+--
+--  * Minimum UTxO 'Value'
+--  * Script transactions require selection of a suitable "collateral" "TxOut".
+--  * Maximum 'TxBody' size.
+--
+--      Usually in bytes, but this module does not prescribe a unit.
+--
+--  * Maximum 'TxOut' size.
+--
+--      For Cardano there are two maximums - one for
+--      serialized size of the 'TxOut' in bytes, and one for the maximum number of
+--      assets in a 'Value'.
+--
+--      Again, this module does not prescribe any particular unit.
+--
+--  * what is reward account balance?
+--
+-- __Optimization__
+--
+-- The coin selection function must minimize costs to the user. The following things should be part of the coin selection function:
+--
+--  * Fee minimization.
+--  * Use the smallest possible number of transactions to satisfy request.
+--  * Over time, in the long run, eventually, utxo distribution should evolve to match payment distribution (emergent utxo management).
+--
+-- __Algorithms__
+--
+--  * v'Sweep' - Select /all/ unspent coins, regardless of transaction outputs.
+--  * v'RandomImprove'
+--  * Random-Improve-MA - Multi-asset variant of Random-Improve.
+--
+
+-- | The transaction type is polymorphic and defined by the caller. Coin
+-- selection however needs to get certain information about the transaction
+-- under construction.
+--
+-- This class defines the types of transaction components and therefore
+-- represents transactions which can have the coin selection algorithms of this
+-- module applied to them.
+class TxBody tx where
```

Comment:
One of the main questions is whether we want concrete types (like `Coin`) or polymorphic types (e.g. `Coin tx`).

* The advantage of concrete types is that the type signatures become simpler — without any loss of generality. For example, the canonical choice for `Amount` is `newtype Amount = Amount Natural`. A coin selection algorithm does not become more general if we use a polymorphic type like `Amount tx` — the type `Natural` embodies exactly the information that we need of a monetary amount.
* However, the advantage of polymorphic types (at least in this context) is that we can plug in existing types without *coercion cost*. If we were to use `newtype Amount = Amount Natural`, then we would have to convert to our `Coin` type, or the `Cardano.Api` equivalent, depending on where we use it. By using `newtype`, we can avoid a *runtime* coercion cost, but we probably still have to pay a *code* coercion cost, as we need to write code that coerces the types.

Our codebase actually contains cautionary tales for both approaches:
* The plethora of type class contexts in `Cardano.Wallet.Primitive.Address*` highlights the complexity of overly polymorphic types.
* `Cardano.Wallet.Shelley.Compatibility` highlights the coercion cost of concrete types.

That said, I lean slightly in favor of not paying the coercing cost, provided that we keep the class `TxBody` in check. For example, that would involve adding constraints such as `Monoid (Amount tx)` to the class definition, and making the type families injective.

Reply:
I don't care much about performance.
But I think polymorphic is better here because then the caller can control the representation, and the unit tests and benchmarks can look simple because they use simple types directly.
`Amount` and `Cost` could reasonably be concreted to `Natural`, but the other types could be anything.